ocular/src/chain/client/cache.rs

```diff
@@ -0,0 +1,519 @@
+
+        // Possible contents is empty, check to avoid parsing errors
+        if !content.is_empty() {
+            toml = match toml::from_str(&content) {
+                Ok(result) => result,
+                Err(err) => {
+                    return Err(CacheError::Toml(err.to_string()));
+                }
+            };
+
+            dbg!(&toml);
+        }
+
+        // Add new item
+        toml.endpoints.push(GrpcEndpoint { address: item });
+
+        let toml_string = toml::to_string(&toml).expect("Could not encode toml value.");
+
+        dbg!(&toml_string);
+
+        // Rewrite file
+        match std::fs::write(&self.path, toml_string) {
+            Ok(_) => Ok(()),
+            Err(err) => Err(CacheError::FileIO(err.to_string())),
+        }
+    }
+
+    fn remove_item(&mut self, item: String) -> Result<(), CacheError> {
+        self.endpoints.remove(&item);
+
+        let mut toml: GrpcEndpointToml = GrpcEndpointToml::default();
+
+        let content = match std::fs::read_to_string(&self.path) {
+            Ok(result) => result,
+            Err(err) => {
+                return Err(CacheError::FileIO(err.to_string()));
+            }
+        };
+
+        // If we were able to get to this point, contents should never be empty, so no need to check.
+        let old_toml: GrpcEndpointToml = match toml::from_str(&content) {
+            Ok(result) => result,
+            Err(err) => {
+                return Err(CacheError::Toml(err.to_string()));
+            }
+        };
+
+        dbg!(&old_toml);
+
+        // Remove item by virtue of excludng it from new toml
+        toml.endpoints = old_toml
+            .endpoints
+            .into_iter()
+            .filter(|ep| ep.address.trim() != item.as_str())
+            .collect();
+
+        let toml_string = toml::to_string(&toml).expect("Could not encode toml value.");
+
+        dbg!(&toml_string);
+
+        // Rewrite file
+        match std::fs::write(&self.path, toml_string) {
+            Ok(_) => Ok(()),
+            Err(err) => Err(CacheError::FileIO(err.to_string())),
+        }
+    }
+
+    fn get_all_items(&self) -> Result<HashSet<String>, CacheError> {
+        Ok(self.endpoints.clone())
+    }
+}
+
+/// Memory based cache
+pub struct MemoryCache {
+    endpoints: HashSet<String>,
+}
+
+impl GrpcCache for MemoryCache {
+    fn is_initialized(&self) -> bool {
+        // No special intialization process so it can always be considered initialized for now.
+        true
+    }
+
+    fn add_item(&mut self, item: String) -> Result<(), CacheError> {
+        self.endpoints.insert(item);
+
+        Ok(())
+    }
+
+    fn remove_item(&mut self, item: String) -> Result<(), CacheError> {
+        self.endpoints.remove(&item);
+
+        Ok(())
+    }
+
+    fn get_all_items(&self) -> Result<HashSet<String>, CacheError> {
+        Ok(self.endpoints.clone())
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
These tests seem involved enough to warrant being in the integration tests dir and run in a docker env since they write to the file system

Reply:
moved to docker integration tests